# 1. Creating dataset with ham and phishing samples

In [ ]:
random_seed = 1337
import pandas as pd 

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


Loading ham samples and creating one dataset from them.

In [ ]:
from joblib import load, dump

df_1 = load('backup_dumps/df_ham_sample_first_10k')
df_2 = load('backup_dumps/df_ham_sample_second_10k')
df_3 = load('backup_dumps/df_ham_sample_third_10k')
df_4_1 = load('backup_dumps/df_ham_sample_fourth_5k_1')
df_4_2 = load('backup_dumps/df_ham_sample_fourth_1k_2')
df_5 = load('backup_dumps/df_ham_sample_fourth_1k_3')
df_6 = load('backup_dumps/df_ham_sample_fifth_20k')
df_7 = load('backup_dumps/df_ham_sample_sixth_20k')
df_8 = load('backup_dumps/df_ham_sample_last')


In [ ]:
df_ham = pd.concat([df_1, df_2, df_3, df_4_1, df_4_2, df_5, df_6, df_7, df_8], ignore_index=True)

In [ ]:
df_ham.head(10)

filename                                         email_body  \
0  enron_extracted.mbox  Andy. Can you put a letter with the facts. As ...   
1  enron_extracted.mbox  Luiz:\n \nI don't know if seating will be limi...   
2  enron_extracted.mbox  Stacey,\nWould I also be able to get this calc...   
3  enron_extracted.mbox  \n\n-----Original Message-----\nFrom: Moses, C...   
4  enron_extracted.mbox  Please be advised that we are in receipt of a ...   
5  enron_extracted.mbox  These are worth a read!!!\n\n> > SEVEN REASONS...   
6  enron_extracted.mbox  ---------------------- Forwarded by Larry F Ca...   
7  enron_extracted.mbox  Good morning.\n\nIn order to finish up Delta/S...   
8  enron_extracted.mbox  CALENDAR ENTRY:\tINVITATION\n\nDescription:\n\...   
9  enron_extracted.mbox  I talked to Keith Holst who went through this ...   

   file_key                                  preprocessed_body  \
0    214081  andy can you put a letter with the facts as a ...   
1    224500  luiz i dont know if seating will be limited bu...   
2    251738  stacey would i also be able to get this calcul...   
3    161746  original message from moses cassy sent thursda...   
4     27457  please be advised that we are in receipt of a ...   
5    200387  these are worth a read seven reasons to crawl ...   
6      8604  forwarded by larry f campbellnaenron on 022820...   
7    150048  good morning in order to finish up deltasalmon...   
8    253272  calendar entry invitation description jennys b...   
9     46611  i talked to keith holst who went through this ...   

                                      tokenized_body  
0  [andy, letter, facts, starter, want, send, let...  
1  [luiz, nt, know, seating, limited, allocated, ...  
2  [stacey, able, calculation, 28th, backed, deal...  
3  [original, message, moses, cassy, sent, thursd...  
4  [advised, receipt, distribution, escrow, agent...  
5  [worth, read, seven, reasons, crawl, rock, 1, ...  
6  [forwarded, larry, f, campbellnaenron, 0228200...  
7  [good, morning, order, finish, deltasalmon, to...  
8  [calendar, entry, invitation, description, jen...  
9  [talked, keith, holst, went, process, couple, ...

In [ ]:
df_ham['is_phishing'] = False

In [ ]:
df_phishing = load('backup_dumps/df_phishing_tokenized')

In [ ]:
df_phishing['is_phishing'] = True

In [ ]:
df_phishing.head(10)

filename                                         email_body  file_key  \
0  20051114.mbox  <html>\n<head>\n<!--   extraneous meta tag rem...         0   
1  20051114.mbox  &nbsp;<body bgcolor=#ffffff><div align="left">...         1   
2  20051114.mbox  &nbsp;<body bgcolor=#ffffff><div align="left">...         2   
3  20051114.mbox  <html><p><font face="Arial"><A HREF="https://w...         3   
4  20051114.mbox  \n\n<xbody bgcolor="#ffffff"><!--Header code s...         4   
5  20051114.mbox  <DIV><DIV><DIV><DIV><DIV><DIV><DIV><DIV><DIV><...         5   
6  20051114.mbox  <DIV><DIV><DIV><DIV><DIV><DIV><DIV><DIV><DIV><...         6   
7  20051114.mbox  <DIV><DIV><DIV><DIV><DIV><DIV><DIV><DIV><DIV><...         7   
8  20051114.mbox  <!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 T...         8   
9  20051114.mbox  To report spam please visit our site at\nhttp:...         9   

                                   preprocessed_body  \
0  ebay suspension fixedstringurl fixedstringurl ...   
1  dear lasalle member as part of our continuing ...   
2  dear lasalle member as part of our continuing ...   
3  fixedstringurl i dont pokemon in 1878 in 1874 ...   
4  fixedstringurl fixedstringurl fixedstringurl f...   
5  fixedstringurl fixedstringurl fixedstringurl f...   
6  fixedstringurl fixedstringurl fixedstringurl f...   
7  fixedstringurl fixedstringurl fixedstringurl f...   
8  untitled document fixedstringurl urgent safeha...   
9  to report spam please visit our site at fixeds...   

                                      tokenized_body  is_phishing  
0  [ebay, suspension, fixedstringurl, fixedstring...         True  
1  [dear, lasalle, member, part, continuing, comm...         True  
2  [dear, lasalle, member, part, continuing, comm...         True  
3  [fixedstringurl, dont, pokemon, 1878, 1874, sh...         True  
4  [fixedstringurl, fixedstringurl, fixedstringur...         True  
5  [fixedstringurl, fixedstringurl, fixedstringur...         True  
6  [fixedstringurl, fixedstringurl, fixedstringur...         True  
7  [fixedstringurl, fixedstringurl, fixedstringur...         True  
8  [untitled, document, fixedstringurl, urgent, s...         True  
9  [report, spam, please, visit, site, fixedstrin...         True

In [ ]:
df_phishing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10569 entries, 0 to 10568
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   filename           10569 non-null  object
 1   email_body         10569 non-null  object
 2   file_key           10569 non-null  int64 
 3   preprocessed_body  10569 non-null  object
 4   tokenized_body     10569 non-null  object
 5   is_phishing        10569 non-null  bool  
dtypes: bool(1), int64(1), object(4)
memory usage: 423.3+ KB


In [ ]:
df_ham.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 96947 entries, 0 to 96946
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   filename           96947 non-null  object
 1   email_body         96947 non-null  object
 2   file_key           96947 non-null  int64 
 3   preprocessed_body  96947 non-null  object
 4   tokenized_body     96947 non-null  object
 5   is_phishing        96947 non-null  bool  
dtypes: bool(1), int64(1), object(4)
memory usage: 3.8+ MB


Creating one dataset from df_ham and df_phishing.

In [ ]:
df_whole_dataset = pd.concat([df_phishing, df_ham], ignore_index=True)

In [ ]:
dump(df_whole_dataset, 'backup_dumps\df_whole_dataset_raw')

['backup_dumps\\df_whole_dataset_raw']

In [ ]:
from joblib import load
df_whole_dataset = load('backup_dumps\df_whole_dataset_raw')

In [ ]:
df_whole_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 107516 entries, 0 to 107515
Data columns (total 6 columns):
 #   Column             Non-Null Count   Dtype 
---  ------             --------------   ----- 
 0   filename           107516 non-null  object
 1   email_body         107516 non-null  object
 2   file_key           107516 non-null  int64 
 3   preprocessed_body  107516 non-null  object
 4   tokenized_body     107516 non-null  object
 5   is_phishing        107516 non-null  bool  
dtypes: bool(1), int64(1), object(4)
memory usage: 4.2+ MB


Data cleanup, removing records that have no tokens in tokenized_body column.

In [ ]:
df_whole_dataset[df_whole_dataset['tokenized_body'].str.len() < 1]

filename             email_body  file_key  \
10844   enron_extracted.mbox       How did it go?\n     59614   
11499   enron_extracted.mbox                 Done\n     22896   
14772   enron_extracted.mbox               Either\n    194637   
17312   enron_extracted.mbox          Here it is!\n     92317   
18017   enron_extracted.mbox     i will be there \n    130891   
...                      ...                    ...       ...   
98338   enron_extracted.mbox  done. see you there\n    131872   
102018  enron_extracted.mbox                   so\n    219497   
103562  enron_extracted.mbox     I will be there.\n    187997   
105902  enron_extracted.mbox          is that it?\n    132625   
107194  enron_extracted.mbox              call me\n     61039   

         preprocessed_body tokenized_body  is_phishing  
10844        how did it go             []        False  
11499                 done             []        False  
14772               either             []        False  
17312           here it is             []        False  
18017      i will be there             []        False  
...                    ...            ...          ...  
98338   done see you there             []        False  
102018                  so             []        False  
103562     i will be there             []        False  
105902          is that it             []        False  
107194             call me             []        False  

[73 rows x 6 columns]

In [ ]:
df_whole_dataset = df_whole_dataset[df_whole_dataset['tokenized_body'].str.len() >= 1]
df_whole_dataset.reset_index(inplace=True, drop=True)

In [ ]:
df_whole_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 107443 entries, 0 to 107442
Data columns (total 6 columns):
 #   Column             Non-Null Count   Dtype 
---  ------             --------------   ----- 
 0   filename           107443 non-null  object
 1   email_body         107443 non-null  object
 2   file_key           107443 non-null  int64 
 3   preprocessed_body  107443 non-null  object
 4   tokenized_body     107443 non-null  object
 5   is_phishing        107443 non-null  bool  
dtypes: bool(1), int64(1), object(4)
memory usage: 4.2+ MB


In [ ]:
from joblib import dump
dump(df_whole_dataset, 'backup_dumps/df_whole_dataset_drop')

['backup_dumps/df_whole_dataset_drop']

# 2. Training the word2vec model to create word embeddings.

We are working with dataset that has 107443 records (around 10k phishing e-mails and 90k ham e-mails). 

We are performing word2vec training on the 'tokenized_body' column that has words from e-mail body, for example sentence: "I will meet you there" is tokenized to a list: ['i', 'will', 'meet', 'you', 'there'].

Word2vec is initialized with options:
* **vector_size=200** - Numbers of dimensions of word vector. Each word will be represented as 200-dimensional vector.
* **window=5** - Maximum distance between the current and predicted word within a sentence (left and right distance). For example, if the target word is "use", in sentence: "We should use word embeddings because there are effective", model will consider words "we", "should", "word", "embeddings", "because", "there", "are" as context. 
* **min_count=6** - Ignores all words with total frequency lower than this.
* **sg=1 ** - Training algorithm, 1 is for Skip-gram model.
* **negative=5** - Number of "negative" samples using during training. These will be the non-context words for each positive (content) word. This technique improves quality of learned embeddings.
* **hs=0 ** - If hs is 0, and negative parameter is non zero, negative sampling will be used.
* **workers=1 ** - Using one thread to train the model (training was done on local machine).
* **seed=random_seed** - Seed to ensure deterministic behaviour.



In [ ]:
from gensim.models import Word2Vec
word2vec_model = Word2Vec(sentences=df_whole_dataset['tokenized_body'], vector_size=200, window=5, min_count=6, sg=1, negative=5, hs=0, workers=1, seed=random_seed)

In [ ]:
word2vec_model.save('models/word2vec_model.model')

In [3]:
import gensim
word2vec_model = gensim.models.Word2Vec.load('/content/drive/MyDrive/phishing_detection_colab/models/word2vec_model.model')

In [7]:
list(word2vec_model.wv.key_to_index.keys())[:20]

['fixedstringemails',
 'fixedstringurl',
 'subject',
 'enron',
 'pm',
 'email',
 'message',
 '2001',
 'sent',
 'original',
 'cc',
 'new',
 'nt',
 'information',
 'energy',
 'power',
 'know',
 'thanks',
 'time',
 'account']

Examples of most similiar words for "click", "phishing", "activate", "expire", "bank". Score is based on: 

In [ ]:
word2vec_model.wv.most_similar('click')

[('link', 0.6355719566345215),
 ('customoffers', 0.5789169073104858),
 ('autowebcom', 0.5774678587913513),
 ('doubleclicking', 0.5704083442687988),
 ('crossdailycom', 0.5683172345161438),
 ('fixedstringurl', 0.5651513934135437),
 ('800com', 0.5649303793907166),
 ('button', 0.5643227100372314),
 ('visitfixedstringurl', 0.5637587904930115),
 ('clickable', 0.560949981212616)]

In [ ]:
word2vec_model.wv.most_similar('phishing')

[('swindlers', 0.7342047095298767),
 ('noncustomers', 0.6466224789619446),
 ('hacking', 0.6243668794631958),
 ('wri', 0.6203393340110779),
 ('bankofthewest', 0.6080707907676697),
 ('lasallebankcom', 0.5865514278411865),
 ('spoof', 0.5831308364868164),
 ('lasallebank', 0.5830559134483337),
 ('gibberish', 0.5760278701782227),
 ('denialofservice', 0.5731956958770752)]

In [ ]:
word2vec_model.wv.most_similar('activate')

[('fixedstringurl7ncuaupdatehtm', 0.5958570241928101),
 ('activation', 0.5531051754951477),
 ('logout', 0.5512038469314575),
 ('complecte', 0.5489070415496826),
 ('unionfcu', 0.5454343557357788),
 ('antitheft', 0.545307993888855),
 ('1fixedstringurl', 0.5421379208564758),
 ('banknet', 0.5395566821098328),
 ('hotlink', 0.5372851490974426),
 ('httpxn11bv1bfkd8grab5ixacompl', 0.5309715270996094)]

In [ ]:
word2vec_model.wv.most_similar('expire')

[('2001does', 0.628576934337616),
 ('expires', 0.6151098608970642),
 ('joseyour', 0.5793957114219666),
 ('unexercised', 0.5545997023582458),
 ('emailbox', 0.5416226983070374),
 ('commences', 0.5274308323860168),
 ('24hours', 0.5267024040222168),
 ('temporal', 0.5261765718460083),
 ('2001and', 0.5251412987709045),
 ('expired', 0.5234308242797852)]

In [ ]:
word2vec_model.wv.most_similar('bank')

[('erste', 0.6764606833457947),
 ('fixedstringurl8005392968the', 0.6652169823646545),
 ('anz', 0.6587746739387512),
 ('fixedstringurl8008414000key', 0.6561096906661987),
 ('canara', 0.6545292735099792),
 ('111000012', 0.6489566564559937),
 ('fixedstringurl8772427837citibank', 0.6477855443954468),
 ('banks', 0.6412453055381775),
 ('fixedstringurl8009723030fleet', 0.6372665166854858),
 ('fixedstringurl8006273999commerce', 0.6331050395965576)]

In [ ]:
len(word2vec_model.wv.index_to_key) # number of words in the vocabulary

72584

# 3. Performing operations with word2vec and final dataset to prepare it for neural network.

Replace tokens with word2vec model index. 

Assign index from word2vec model. 

Assign **special_index** preventing Out Of Vocabulary (if word is not present in the model's vocabulary, we must assign it an index).

In [5]:
def tokens_to_indices(model, list_of_tokens, special_index):
    tokens_indices = [model.wv.key_to_index.get(token, special_index) for token in list_of_tokens]
    return tokens_indices

In [5]:
vocab_len = len(word2vec_model.wv.index_to_key)

In [6]:
vocab_len

72584

In [9]:
special_index = vocab_len + 10000

In [12]:
print(tokens_to_indices(word2vec_model, ['probablynotfoundtoken', 'probablynotfoundtoken2', 'phishing'], special_index))

[82584, 82584, 12290]


In [2]:
from joblib import load
df_whole_dataset_drop = load('/content/drive/MyDrive/phishing_detection_colab/datasets/df_whole_dataset_drop')

In [2]:
df_whole_dataset_drop.head(10)

filename                                         email_body  file_key  \
0  20051114.mbox  <html>\n<head>\n<!--   extraneous meta tag rem...         0   
1  20051114.mbox  &nbsp;<body bgcolor=#ffffff><div align="left">...         1   
2  20051114.mbox  &nbsp;<body bgcolor=#ffffff><div align="left">...         2   
3  20051114.mbox  <html><p><font face="Arial"><A HREF="https://w...         3   
4  20051114.mbox  \n\n<xbody bgcolor="#ffffff"><!--Header code s...         4   
5  20051114.mbox  <DIV><DIV><DIV><DIV><DIV><DIV><DIV><DIV><DIV><...         5   
6  20051114.mbox  <DIV><DIV><DIV><DIV><DIV><DIV><DIV><DIV><DIV><...         6   
7  20051114.mbox  <DIV><DIV><DIV><DIV><DIV><DIV><DIV><DIV><DIV><...         7   
8  20051114.mbox  <!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 T...         8   
9  20051114.mbox  To report spam please visit our site at\nhttp:...         9   

                                   preprocessed_body  \
0  ebay suspension fixedstringurl fixedstringurl ...   
1  dear lasalle member as part of our continuing ...   
2  dear lasalle member as part of our continuing ...   
3  fixedstringurl i dont pokemon in 1878 in 1874 ...   
4  fixedstringurl fixedstringurl fixedstringurl f...   
5  fixedstringurl fixedstringurl fixedstringurl f...   
6  fixedstringurl fixedstringurl fixedstringurl f...   
7  fixedstringurl fixedstringurl fixedstringurl f...   
8  untitled document fixedstringurl urgent safeha...   
9  to report spam please visit our site at fixeds...   

                                      tokenized_body  is_phishing  
0  [ebay, suspension, fixedstringurl, fixedstring...         True  
1  [dear, lasalle, member, part, continuing, comm...         True  
2  [dear, lasalle, member, part, continuing, comm...         True  
3  [fixedstringurl, dont, pokemon, 1878, 1874, sh...         True  
4  [fixedstringurl, fixedstringurl, fixedstringur...         True  
5  [fixedstringurl, fixedstringurl, fixedstringur...         True  
6  [fixedstringurl, fixedstringurl, fixedstringur...         True  
7  [fixedstringurl, fixedstringurl, fixedstringur...         True  
8  [untitled, document, fixedstringurl, urgent, s...         True  
9  [report, spam, please, visit, site, fixedstrin...         True

In [11]:
df_whole_dataset_drop['tokenized_body_indexed'] = df_whole_dataset_drop['tokenized_body'].apply(lambda x: tokens_to_indices(word2vec_model, x, special_index))

In [23]:
df_whole_dataset_drop.head(10)

filename                                         email_body  file_key  \
0  20051114.mbox  <html>\n<head>\n<!--   extraneous meta tag rem...         0   
1  20051114.mbox  &nbsp;<body bgcolor=#ffffff><div align="left">...         1   
2  20051114.mbox  &nbsp;<body bgcolor=#ffffff><div align="left">...         2   
3  20051114.mbox  <html><p><font face="Arial"><A HREF="https://w...         3   
4  20051114.mbox  \n\n<xbody bgcolor="#ffffff"><!--Header code s...         4   
5  20051114.mbox  <DIV><DIV><DIV><DIV><DIV><DIV><DIV><DIV><DIV><...         5   
6  20051114.mbox  <DIV><DIV><DIV><DIV><DIV><DIV><DIV><DIV><DIV><...         6   
7  20051114.mbox  <DIV><DIV><DIV><DIV><DIV><DIV><DIV><DIV><DIV><...         7   
8  20051114.mbox  <!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 T...         8   
9  20051114.mbox  To report spam please visit our site at\nhttp:...         9   

                                   preprocessed_body  \
0  ebay suspension fixedstringurl fixedstringurl ...   
1  dear lasalle member as part of our continuing ...   
2  dear lasalle member as part of our continuing ...   
3  fixedstringurl i dont pokemon in 1878 in 1874 ...   
4  fixedstringurl fixedstringurl fixedstringurl f...   
5  fixedstringurl fixedstringurl fixedstringurl f...   
6  fixedstringurl fixedstringurl fixedstringurl f...   
7  fixedstringurl fixedstringurl fixedstringurl f...   
8  untitled document fixedstringurl urgent safeha...   
9  to report spam please visit our site at fixeds...   

                                      tokenized_body  is_phishing  \
0  [ebay, suspension, fixedstringurl, fixedstring...         True   
1  [dear, lasalle, member, part, continuing, comm...         True   
2  [dear, lasalle, member, part, continuing, comm...         True   
3  [fixedstringurl, dont, pokemon, 1878, 1874, sh...         True   
4  [fixedstringurl, fixedstringurl, fixedstringur...         True   
5  [fixedstringurl, fixedstringurl, fixedstringur...         True   
6  [fixedstringurl, fixedstringurl, fixedstringur...         True   
7  [fixedstringurl, fixedstringurl, fixedstringur...         True   
8  [untitled, document, fixedstringurl, urgent, s...         True   
9  [report, spam, please, visit, site, fixedstrin...         True   

                              tokenized_body_indexed  
0  [111, 1851, 1, 1, 1, 1, 1, 111, 1851, 1, 1, 1,...  
1  [169, 9578, 249, 2154, 1583, 1629, 537, 19, 11...  
2  [169, 9578, 249, 2154, 1583, 1629, 537, 19, 11...  
3                [1, 3047, 82584, 51669, 32874, 534]  
4                                 [1, 1, 1, 1, 1, 1]  
5  [1, 1, 1, 1, 1592, 141, 1, 1, 1, 1, 1, 4517, 1...  
6  [1, 1, 1, 1, 1592, 141, 1, 1, 1, 1, 1, 4517, 1...  
7  [1, 1, 1, 1, 1592, 141, 1, 1, 1, 1, 1, 4517, 1...  
8  [5471, 392, 1, 3695, 9420, 306, 256, 111, 1197...  
9  [80, 4031, 118, 164, 226, 1, 5, 856, 93, 600, ...

In [24]:
df_whole_dataset_drop.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 107443 entries, 0 to 107442
Data columns (total 7 columns):
 #   Column                  Non-Null Count   Dtype 
---  ------                  --------------   ----- 
 0   filename                107443 non-null  object
 1   email_body              107443 non-null  object
 2   file_key                107443 non-null  int64 
 3   preprocessed_body       107443 non-null  object
 4   tokenized_body          107443 non-null  object
 5   is_phishing             107443 non-null  bool  
 6   tokenized_body_indexed  107443 non-null  object
dtypes: bool(1), int64(1), object(5)
memory usage: 5.0+ MB


Saving the dataset to disk and then loading it.

In [12]:
from joblib import dump
dump(df_whole_dataset_drop, '/content/drive/MyDrive/phishing_detection_colab/datasets/df_whole_dataset_drop')

['/content/drive/MyDrive/phishing_detection_colab/datasets/df_whole_dataset_drop']

In [4]:
from joblib import load
df_whole_dataset_drop = load('/content/drive/MyDrive/phishing_detection_colab/datasets/df_whole_dataset_drop')

Pad sequences to gain fixed size.

We take the maximum length of tokens in e-mail body (max_tokens_len) and pad indexes of word2vec (tokenized_body_indexed) to ensure fixed size of all sequences.

We use value=vocab_len which is 72584 (word2vec vocab size), to ensure that padded value won't appear in word2vec indices.

In [3]:
max_tokens_len = df_whole_dataset_drop['tokenized_body'].apply(len).max()

In [4]:
max_tokens_len 

55231

In [ ]:
from tensorflow.keras.preprocessing.sequence import pad_sequences

padded_sequences = pad_sequences(df_whole_dataset_drop['tokenized_body_indexed'], maxlen=max_tokens_len, padding='post', value=vocab_len)

In [ ]:
padded_sequences

array([[  111,  1851,     1, ..., 72584, 72584, 72584],
       [  169,  9578,   249, ..., 72584, 72584, 72584],
       [  169,  9578,   249, ..., 72584, 72584, 72584],
       ...,
       [ 3920,   321,  8045, ..., 72584, 72584, 72584],
       [ 2581,  2793,     5, ..., 72584, 72584, 72584],
       [  414,   128,   220, ..., 72584, 72584, 72584]], dtype=int32)

In [ ]:
import numpy as np
np.savez_compressed('/content/drive/MyDrive/phishing_detection_colab/datasets/padded_sequences.npz', my_array=padded_sequences)

Loading below data consumes about 20 GB of RAM.

In [5]:
import numpy as np
loaded_data = np.load('/content/drive/MyDrive/phishing_detection_colab/datasets/padded_sequences.npz')
padded_sequences= loaded_data['my_array']

In [17]:
padded_sequences

array([[  111,  1851,     1, ..., 72584, 72584, 72584],
       [  169,  9578,   249, ..., 72584, 72584, 72584],
       [  169,  9578,   249, ..., 72584, 72584, 72584],
       ...,
       [ 3920,   321,  8045, ..., 72584, 72584, 72584],
       [ 2581,  2793,     5, ..., 72584, 72584, 72584],
       [  414,   128,   220, ..., 72584, 72584, 72584]], dtype=int32)

In [16]:
padded_sequences.size

5934184333

# 4. Creating Bi-LSTM neural network.

Initialize embedding matrix and embedding layer.

In [8]:
import numpy as np

embedding_dim = word2vec_model.vector_size
embedding_matrix = np.zeros((vocab_len, embedding_dim))

for word, idx in word2vec_model.wv.key_to_index.items():
  embedding_matrix[idx] = word2vec_model.wv[word]

In [20]:
embedding_dim

200

In [21]:
np.savez_compressed('/content/drive/MyDrive/phishing_detection_colab/datasets/embedding_matrix.npz', my_array=embedding_matrix)

Split dataset into train and test, about 46 GB of RAM used.

In [6]:
random_seed = 1337
from sklearn.model_selection import train_test_split

X = padded_sequences
y = df_whole_dataset_drop['is_phishing'].values

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=random_seed)

Delete variables from memory to free up RAM memory and save train and test data.

In [7]:
import gc
del X
del y
del padded_sequences
del loaded_data
del df_whole_dataset_drop
del word2vec_model
gc.collect()

0

Above operations clear up about 20 GB of RAM memory.

In [8]:
import tensorflow as tf
train_data = tf.data.Dataset.from_tensor_slices((X_train, y_train))

Perform freeing memory as well, after below operations, about 26.9 GB of RAM is used.

In [9]:
del X_train
del y_train
gc.collect()

41

In [10]:
batch_size = 32
buffer_size = 10000
train_data = train_data.shuffle(buffer_size).batch(batch_size).prefetch(buffer_size)

With using the deleting variable and running garbage collection technique, train data is saved succesfully (takes about 40 GB of RAM)

In [11]:
train_data.save('/content/drive/MyDrive/phishing_detection_colab/datasets/train_data_not_corrupted')

Now delete the train_data and perform same operations on test_data.

In [12]:
del train_data
gc.collect()

0

In [13]:
test_data = tf.data.Dataset.from_tensor_slices((X_test, y_test))

In [14]:
del X_test
del y_test
gc.collect()

0

In [15]:
test_data = test_data.batch(batch_size).prefetch(buffer_size)

In [16]:
test_data.save('/content/drive/MyDrive/phishing_detection_colab/datasets/test_data_not_corrupted')

## Loading all inputs to neural network and performing training.

Now load the train_data and test_data only as long with other variables used in embedding layer.

Finally we have a dataset of 107443 rows (each row is one e-mail) and 55231 columns (each column represent word used as embedding)

Initalize models.


In [6]:
import numpy as np
vocab_len=72584 # word2vec vocabulary size
embedding_dim=200 # dimensionality of word vector
max_tokens_len=55231 # maximum number of tokens 
embedding_matrix_load=np.load('/content/drive/MyDrive/phishing_detection_colab/datasets/embedding_matrix.npz') 
embedding_matrix = embedding_matrix_load['my_array'] # embedding matrix

Below, train_data and test_data takes about 1 GB of RAM.

In [3]:
import tensorflow as tf
train_data=tf.data.Dataset.load('/content/drive/MyDrive/phishing_detection_colab/datasets/train_data_not_corrupted')

In [18]:
test_data=tf.data.Dataset.load('/content/drive/MyDrive/phishing_detection_colab/datasets/test_data_not_corrupted')

In [12]:
train_data

<_LoadDataset element_spec=(TensorSpec(shape=(None, 55231), dtype=tf.int32, name=None), TensorSpec(shape=(None,), dtype=tf.bool, name=None))>

In [13]:
test_data

<_LoadDataset element_spec=(TensorSpec(shape=(None, 55231), dtype=tf.int32, name=None), TensorSpec(shape=(None,), dtype=tf.bool, name=None))>

In [7]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, Bidirectional, LSTM, Dense, Dropout

model = Sequential([
    Embedding(input_dim=vocab_len, output_dim=embedding_dim, input_length=max_tokens_len, weights=[embedding_matrix], trainable=False),
    Bidirectional(LSTM(units=128)),
    Dropout(0.2),
    Dense(1, activation='sigmoid')
])


In [ ]:
!pip install tf-nightly

In [1]:
import tensorflow as tf
precision_metric = tf.keras.metrics.Precision(name='precision')
recall_metric = tf.keras.metrics.Recall(name='recall')
f1_metric = tf.keras.metrics.F1Score(threshold=0.5)

In [14]:
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=[precision_metric, recall_metric])

10 gb of ram,
33.5 gb of gpu
is used

In [ ]:
%%time
model.fit(train_data, epochs=10, validation_data=test_data)

Epoch 1/10
2687/2687 [==============================] - 7995s 3s/step - loss: 0.0342 - precision: 0.9628 - recall: 0.9164 - val_loss: 0.0142 - val_precision: 0.9921 - val_recall: 0.9594
Epoch 2/10
2687/2687 [==============================] - 8008s 3s/step - loss: 0.0236 - precision: 0.9811 - recall: 0.9462 - val_loss: 0.0206 - val_precision: 0.9971 - val_recall: 0.9374
Epoch 3/10
2395/2687 [=========================>....] - ETA: 13:09 - loss: 0.0166 - precision: 0.9867 - recall: 0.9573

In [ ]:
test_loss, test_precision, test_recall = model.evaluate(test_data)
print("Test loss:", test_loss)
print("Test precision:", test_precision)
print("Test recall:", test_recall)


In [ ]:
!apt-get install texlive texlive-xetex texlive-latex-extra pandoc
!pip install pypandoc

In [15]:
!cp drive/MyDrive/ColabNotebooks/Kopianotatnikamodel.ipynb ./

In [17]:
!jupyter nbconvert --to pdf "Kopianotatnikamodel.ipynb"


[NbConvertApp] Converting notebook Kopianotatnikamodel.ipynb to pdf
[NbConvertApp] Writing 89173 bytes to notebook.tex
[NbConvertApp] Building PDF
[NbConvertApp] Running xelatex 3 times: ['xelatex', 'notebook.tex', '-quiet']
[NbConvertApp] Running bibtex 1 time: ['bibtex', 'notebook']
[NbConvertApp] WARNING | bibtex had problems, most likely because there were no citations
[NbConvertApp] PDF successfully created
[NbConvertApp] Writing 85562 bytes to Kopianotatnikamodel.pdf
